### PR TITLE
Provide the basic testing tools for the operator developers

### PR DESCRIPTION
**Pull request by nolar**
> Issue : #52 

Provide the basic tools for the operator developers to test their Kopf-based operators.

This feature will be used also to run end2end tests in the following PRs (e.g. #54).

## Checklist

- [x] Tests
- [x] Documentation

### CODE REVIEW

File: kopf/testing/runner.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+        with KopfRunner(['run', '--verbose', 'examples/01-minimal/example.py']) as runner:
+            # do something while the operator is running.
+            time.sleep(3)
+
+        assert runner.exit_code == 0
+        assert runner.exception is None
+        assert 'And here we are!' in runner.stdout
+
+    All the args & kwargs are passed directly to the Click's invocation method.
```
Comment (samurang87) [inline kopf/testing/runner.py]: "passed directly to Click's..."

File: kopf/testing/runner.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    See: `click.testing.CliRunner`.
+    All properties proxy directly to the Click's `click.testing.Result` object
```
Comment (samurang87) [inline kopf/testing/runner.py]: "proxy directly to Click's `click.testing.Result`"...

File: kopf/testing/runner.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    when it is available (i.e. after the context manager exits).
+
+    CLI commands have to be invoked in parallel threads, never in processes:
+
+    First, with multiprocessing, they are unable to pickle and pass
+    the exceptions (specifically, their traceback objects)
```
Comment (samurang87) [inline kopf/testing/runner.py]: "pickle and pass exceptions..."

File: kopf/testing/runner.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    from a child thread (Kopf's CLI) to the parent thread (pytest).
+
+    Second, mocking works within one process (all threads),
+    but not across the processes --- the mock's calls (counts, arrgs) are lost.
```
Comment (samurang87) [inline kopf/testing/runner.py]: "not across processes..."

File: kopf/testing/runner.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+import threading
+
+import click.testing
+
+from kopf.cli import main
+
+
+class KopfRunner:
+    """
+    A context manager to run Kopf-based operator in parallel with the tests.
```
Comment (samurang87) [inline kopf/testing/runner.py]: "run a Kopf-based operator" or "run Kopf-based operators"

**Comment by nolar** [comment]
@samurang87 All grammar issues are fixed.

File: tests/testing/test_runner.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    assert runner.exit_code == 0
+    assert runner.exception is None
+    assert runner.output.startswith("Usage:")
+    assert runner.stdout.startswith("Usage:")
+    assert runner.stdout_bytes.startswith(b"Usage:")
+    # Note: stderr is not captured, it is mixed with stdout.
+
+
+def test_exception_from_runner_suppressed_with_no_reraise():
+    with KopfRunner(['run', 'non-existant.py', '--standalone'], reraise=False) as runner:
```
Comment (samurang87) [inline tests/testing/test_runner.py]: "existent"

**Comment by nolar** [comment]
@samurang87 Fixed again.

**Comment by samurang87** [comment]
"existant" is in a few places actually :)

**Comment by nolar** [comment]
@samurang87 Fixed in few places.

**Pull request merged**